lib/MenuItem.php

```diff
@@ -510,7 +527,11 @@ public function path() {
 	 */
 	public function title() {
 		if ( isset($this->__title) ) {
-			return $this->__title;
+			/**
+			 * @see Walker_Nav_Menu::start_el()
+			 */
+			$title = apply_filters( 'nav_menu_item_title', $this->__title, $this->core_object, $this->menu->args ? $this->menu->args : new \stdClass, $this->level );
```

Comment:
Nice 😎

Reply:
I have mixed feelings about this and some parts of https://github.com/timber/timber/issues/1820#issue-370922635. 

Timber is actually trying to get rid of those filters everywhere to handle HTML. And menu are a noticeably awful example of this practice. I think we shouldn't put too much effort in being 100% compliant with all those filters (`wp_nav_menu_items`, `wp_nav_menu_{$menu->slug}_items`, etc.). It might not be possible or it might come at too much code costs. 

If some user wants full compatibility with all core menu filters, skipping Timber implemented menus and using `wp_nav_menu` is quite easy. I actually had an idea for this that I didn't commit in that PR that could be a fair compromise between all Timber menus VS `wp_nav_menu` only.

A `__toString` method on the `Menu` class that will dump the HTML menu using walker and all those core stuff:

```php
public function __toString() {
	static $menu_id_slugs = array();

	$args = $this->args;

	$items = '';
	$nav_menu = '';
	$show_container = false;

	if ( $args->container ) {
		/**
			* Filters the list of HTML tags that are valid for use as menu containers.
			*
			* @since 3.0.0
			*
			* @param string[] $tags The acceptable HTML tags for use as menu containers.
			*                       Default is array containing 'div' and 'nav'.
			*/
		$allowed_tags = apply_filters( 'wp_nav_menu_container_allowedtags', array( 'div', 'nav' ) );

		if ( is_string( $args->container ) && in_array( $args->container, $allowed_tags, true ) ) {
			$show_container = true;
			$class          = $args->container_class ? ' class="' . esc_attr( $args->container_class ) . '"' : ' class="menu-' . $this->slug . '-container"';
			$id             = $args->container_id ? ' id="' . esc_attr( $args->container_id ) . '"' : '';
			$aria_label     = ( 'nav' === $args->container && $args->container_aria_label ) ? ' aria-label="' . esc_attr( $args->container_aria_label ) . '"' : '';
			$nav_menu      .= '<' . $args->container . $id . $class . $aria_label . '>';
		}
	}

	$items .= walk_nav_menu_tree( $this->sorted_menu_items, $args->depth, $args );

	// Attributes.
	if ( ! empty( $args->menu_id ) ) {
		$wrap_id = $args->menu_id;
	} else {
		$wrap_id = 'menu-' . $this->slug;

		while ( in_array( $wrap_id, $menu_id_slugs, true ) ) {
			if ( preg_match( '#-(\d+)$#', $wrap_id, $matches ) ) {
				$wrap_id = preg_replace( '#-(\d+)$#', '-' . ++$matches[1], $wrap_id );
			} else {
				$wrap_id = $wrap_id . '-1';
			}
		}
	}
	$menu_id_slugs[] = $wrap_id;

	$wrap_class = $args->menu_class ? $args->menu_class : '';

	$nav_menu .= sprintf( $args->items_wrap, esc_attr( $wrap_id ), esc_attr( $wrap_class ), $items );
	if ( $show_container ) {
		$nav_menu .= '</' . $args->container . '>';
	}

	return $nav_menu;
}
```

So you could simply do `{{ header_menu }}` in your Twig template while still keeping the ability to use `Menu` objects/factories. 

Let me know if think this could be handy.

Reply:
That sounds like a great idea! I just wouldn’t use it with the `__toString()` method, but maybe add a `wp_nav_menu()` method?

```twig
{{ header_menu.wp_nav_menu }}
```

Reply:
I stand by the `__toString` method which makes more sense for the following reasons:
- The HTML menu is literally the string representation of the menu object
- The name `wp_nav_menu` is misleading, users could think it could take the same arguments than the native `wp_nav_menu` WordPress method, which is not

Reply:
@nlemoine I see. Yeah, that makes sense. Let’s go with `__toString()` then 🙂.